store/postgres/src/relational_queries.rs

```diff
@@ -2171,6 +2203,26 @@ impl<'a> FilterCollection<'a> {
             FilterCollection::MultiWindow(windows, _) => windows.is_empty(),
         }
     }
+
+    /// Return the id type of the fields in the parents for which the query
+    /// produces children. This is `None` if there are no parents, i.e., for
+    /// a toplevel query.
+    pub(crate) fn parent_type(&self) -> Result<Option<IdType>, StoreError> {
+        match self {
+            FilterCollection::All(_) => Ok(None),
+            FilterCollection::SingleWindow(window) => Ok(Some(window.parent_type())),
+            FilterCollection::MultiWindow(windows, _) => {
+                let parent_types: HashSet<_> =
+                    HashSet::from_iter(windows.into_iter().map(|window| window.parent_type()));
+                if parent_types.len() > 1 {
+                    return Err(graph::constraint_violation!(
+                        "all implementors of an interface must use the same type for their `id`"
+                    ));
+                }
+                Ok(parent_types.iter().next().map(|t| t.to_owned()))
```

Comment:
We could also use `all_equal()`, as a possibly clearer alternative to the `HashSet` and `> 1` logic:

```suggestion
                if windows.iter().map(FilterWindow::parent_type).all_equal() {
                    Ok(Some(windows[0].parent_type().to_owned()))
                } else {
                    Err(graph::constraint_violation!(
                        "all implementors of an interface must use the same type for their `id`"
                    ))
                }
```

Just food for thought; feel free to ignore this.

Reply:
I didn't know about `all_equal`. I like your way a lot better. Changed.